### PR TITLE
Crop rotation fixes

### LINKED ISSUE #324: rotate and crop errors
From https://www.openmicroscopy.org/community/posting.php?mode=quote&f=4&p=20674
A user here at HMS reported to me that they were having problems with rotated and cropped figures. I was able to reproduce the problem. When you crop a figure that has been rotated the position of the cropping is not consistent with the output you get. The attached screenshot illustrates what is happening. The Preview pane shows what is intended. The resulting figure is offset to the lower left. The problem seems to be in the accuracy of the crop tool. If you reposition the crop frame only slightly, even though it appears to be in the wrong place, the correct crop is possible.

### LINKED ISSUE #408: ROI rectangle inset rotation
See https://forum.image.sc/t/omero-figure-box-annotations-rotated-incorrectly/45787

Copy the crop region from a rotated image and paste on it's "parent" image - but the new Rectangle isn't rotated.

Just playing with css tools, a fix can be simulated with a transform of negative rotation and transform-origin. See:

Screenshot 2020-11-26 at 10 27 43

But, we'd need to save this and handle it somehow when Editing the rectangle?

### PR DESCRIPTION
**Pull request by will-moore**
Fixes #408 - copy rotated crop region and paste to ROI.
Also fixes #324

Rectangle ROIs don't support rotation, but this fixes the issue above by creating a Polygon to simulate a rotated Rectangle.

To test:
 - Copy & Paste to duplicate a panel, then zoom in to crop, and rotate one of them.
 - Copy the Crop region and Paste it as a new Shape on the un-zoomed panel - The new Shape (Polygon) should match the crop region:

Screenshot 2020-12-08 at 21 55 42

 - Copy the rotated Crop region, select the uncropped (unrotated) Panel and click Paste to paste the rotated crop region. The rotation should also be applied to the panel so that both images show the same rotated viewport.
 - Crop the rotated panel using the "Crop" button. The 'initial' crop region should match the current viewport (see screenshot) and the updated crop region should be correctly applied to the panel.

Screenshot 2020-12-15 at 14 28 10

### CODE REVIEW

**Comment by pwalczysko** [comment]
> Copy the rotated Crop region, select the uncropped (unrotated) Panel and click Paste to paste the rotated crop region. The rotation should also be applied to the panel so that both images show the same rotated viewport.

Up till this point all worked. The above one I am not sure what to do with - where is the unrotated uncropped Panel ? I just used it in the previous step, and pasted the ROI on it, and it was rotated.... as I would expect.

**Comment by will-moore** [comment]
> Copy the rotated Crop region, select the uncropped (unrotated) Panel and click Paste to paste the rotated crop region.

This means to use the "Paste" button at the bottom of the "Preview" Tab, to apply the same crop and rotation to the selected panel. This is NOT Pasting to create a new ROI as in previous steps.

**Comment by pwalczysko** [comment]
> > Copy the rotated Crop region, select the uncropped (unrotated) Panel and click Paste to paste the rotated crop region.
> 
> This means to use the "Paste" button at the bottom of the "Preview" Tab, to apply the same crop and rotation to the selected panel. This is NOT Pasting to create a new ROI as in previous steps.

Thanks, got this one, and it works as expected.

**Comment by pwalczysko** [comment]
> Crop the rotated panel using the "Crop" button. The 'initial' crop region should match the current viewport (see screenshot) and the updated crop region should be correctly applied to the panel.

Sorry, struggling with this one too. Using the Crop button is clear, but should I first rotate and then Crop ? This means I should not zoom first ? Should I copy and paste the ROI somewhere ?  If I do as told here, I am ending up with see screenshot below, how do I get the ROI displayed on the rotated image as shown on yours ?

Screenshot 2021-01-11 at 18 05 12

**Comment by pwalczysko** [comment]
> Crop the rotated panel using the "Crop" button. The 'initial' crop region should match the current viewport (see screenshot) and the updated crop region should be correctly applied to the panel.

After clarification, I was able to verify the above workflow too, it works fine. 

Nevertheless, in this workflow, if the ROI does not come from the zoomed-in region but, is copied from another place in Figure (screenshot 1 below) or comes from OMERO (screenshot 2 below), this workflow still fails, as the copied or "from OMERO" ROIs are displayed unrotated.

Screenshot 1 (ROI copied from somewhere else in Figure and pasted (option "From Clipboard")

Screenshot 2021-01-12 at 11 17 17

Screenshot 2 (ROI from OMERO in Crop dialog, ROI should be rotated as the image is rotated)

Screenshot 2021-01-12 at 11 18 19

**Comment by pwalczysko** [comment]
Would it be possible to fix also the Edit ROI panel view for rotated images ?

Workflow: Rotate an image in Figure. Click on "Labels" tab and "Edit" in the "ROI" row. Observe that the view in the Edit ROI dialog is unrotated, the image inside the panel on the canvas is rotated though.

Screenshot 2021-01-12 at 11 25 02

**Comment by pwalczysko** [comment]
In summary, 3 RFEs : 2 RFEs regarding the crop dialog and rotation https://github.com/ome/omero-figure/pull/410#issuecomment-758591609 and one regarding the Edit ROI dialog https://github.com/ome/omero-figure/pull/410#issuecomment-758593581

**Comment by will-moore** [comment]
@pwalczysko I'm afraid fixing the "Edit ROIs" dialog to handle rotated images isn't easy since we'd need drawing tools to allow you to draw rotated Rectangles which I don't have (hand-coded drawing tools).

But the other 2 RFEs above should be fixed:
 - Open the crop dialog, having either copied a Crop Region from another Panel (with or without rotation) OR having copied a Rectangle. Also nice to have Rectangle ROIs saved on the Image in OMERO, and/or Rectangle ROIs on the Figure panel (see screenshot)
 - Check that each of these thumbnails shows the correct region of the image.
 - Clicking either the Rectangle ROIs from OMERO or the Rectangle on the figure panel (top and bottom sections of thumbnails) will set the Panel rotation to zero. Clicking the "clipboard" thumbnail will apply a rotation IF you copied the crop region from a Panel that was rotated.
 - When you hit "OK", any rotation that got applied should be saved (so that the crop region you selected was correctly applied to the image).

Screenshot 2021-01-22 at 10 36 09

**Comment by pwalczysko** [comment]
> @pwalczysko I'm afraid fixing the "Edit ROIs" dialog to handle rotated images isn't easy since we'd need drawing tools to allow you to draw rotated Rectangles which I don't have (hand-coded drawing tools).

I think this is fine. I would suggest a (couple of) short video(s) for the solution of the rotation and crop workflows would be very handy, as these workflows are very popular, but as the testing shows, not easy to get right on the first try.

Would you like to list for testing tomorrow with my name ?

**Comment by pwalczysko** [comment]
> But the other 2 RFEs above should be fixed:

    Open the crop dialog, having either copied a Crop Region from another Panel (with or without rotation) OR having copied a Rectangle. Also nice to have Rectangle ROIs saved on the Image in OMERO, and/or Rectangle ROIs on the Figure panel (see screenshot)
    Check that each of these thumbnails shows the correct region of the image.
    Clicking either the Rectangle ROIs from OMERO or the Rectangle on the figure panel (top and bottom sections of thumbnails) will set the Panel rotation to zero. Clicking the "clipboard" thumbnail will apply a rotation IF you copied the crop region from a Panel that was rotated.
    When you hit "OK", any rotation that got applied should be saved (so that the crop region you selected was correctly applied to the image).

This works as expected. Ready to merge fmpov.

**Comment by will-moore** [comment]
I made a demo movie for this PR:

https://user-images.githubusercontent.com/900055/106256644-731b4600-6213-11eb-90d1-e3c4173590b9.mp4

**Comment by pwalczysko** [comment]
@will-moore great movie, thanks, these are precisely the workflows which were picked upon in the recent workshop again, and the difference of those two Paste buttons is nicely shown

**Comment by jburel** [comment]
This could probably go to the official youtube channel otherwise it will be lost

**Comment by will-moore** [comment]
Yes, can make the video more "public" when we release.
The last commit needs a test before merging: Copy the crop region from a non-rotated image (zero rotation) and paste as a crop region onto a rotated image. The rotation should be set to zero.

**Comment by pwalczysko** [comment]
> Copy the crop region from a non-rotated image (zero rotation) and paste as a crop region onto a rotated image. The rotation should be set to zero.

Tested
When I use the Paste buttion from Labels pane, the rotation is not set to zero. When I use the Paste button from Preview pane, the rotation is not set to zero. I guess this is the expected behaviour ?

**Comment by will-moore** [comment]
> When I use the Paste button from Preview pane, the rotation is not set to zero. I guess this is the expected behaviour ?

No. When I use the Paste button from Preview pane, the rotation should be set to zero.

**Comment by pwalczysko** [comment]
> > When I use the Paste button from Preview pane, the rotation is not set to zero. I guess this is the expected behaviour ?
> 
> No. When I use the Paste button from Preview pane, the rotation should be set to zero.

Yes, the rotation on Pate is indeed set to zero. Sorry, typo (in my original report, there are two "no"s for both of the panes, in truth, I shoiuld have written "y" for the Preview pane, as this is what I saw.

Sorry

File: src/js/views/crop_modal_view.js [hunk 1]
```diff
@@ -167,6 +179,18 @@ var CropModalView = Backbone.View.extend({
                 if (sameT) {
                     t = self.m.get('theT');
                 }
+
+                var rotation = self.m.get('rotation');
+                if (rotation != 0) {
```
Comment (jburel) [inline src/js/views/crop_modal_view.js:184]: could that section of code be turned into a function?
Same code above using ``roi`` instead of ``r``.

Comment (will-moore) [reply src/js/views/crop_modal_view.js]: Fixed in 88cafb4

**Comment by pwalczysko** [comment]
Inside the Crop dialog, copied a Rrectangle ROI from another panel which was rotated so that it is in Clipboard now.

the desciption above says

> Clicking the "clipboard" thumbnail will apply a rotation IF you copied the crop region from a Panel that was rotated.

But when I click the Clipboard thumb it does not apply any rotation, instead, it unrotates the main panel in the Crop dialog and pastes the ROI correctly. Did this behaviour change (i.e. unitended change) or is it a misunderstanding ?

**Comment by will-moore** [comment]
You copied a Rectangle ROI, but my text says "IF you copied the *crop region*..."
Rectangles can't have any rotation, but a crop region can.
The behaviour hasn't changed.

Does the thumbnail representation that you clicked look like the crop outline that is generated when you click it?

**Comment by pwalczysko** [comment]
> You copied a Rectangle ROI, but my text says "IF you copied the _crop region_..."
> Rectangles can't have any rotation, but a crop region can.
> The behaviour hasn't changed.
> 
> Does the thumbnail representation that you clicked look like the crop outline that is generated when you click it?

thanks for explaining. 
No, the crop outline looks differently, the crop dialog is not showing the right thing now imho when Clipboard is selected.

Screenshot 2021-02-09 at 11 45 08

**Comment by pwalczysko** [comment]
Compare the situation inside Crop dialog described in previous comment with the situation when I paste the crop region without using the Crop dialog on the same image.

Screenshot 2021-02-09 at 11 48 20

This behaviour seems right here, whereas the Crop dialog above https://github.com/ome/omero-figure/pull/410#issuecomment-775880924 shows the wrong thumb and when thumb is clicked, a wrong rotation and position is shown in main panel.

**Comment by will-moore** [comment]
Sorry @pwalczysko I can't seem to reproduce the bug you're seeing.
Can you point me at an example figure and exact steps to reproduce? Thanks.

**Comment by pwalczysko** [comment]
The image used is

https://merge-ci.openmicroscopy.org/web/webclient/?show=image-140172

The figure saved is

https://merge-ci.openmicroscopy.org/web/figure/file/70451

We clarified the workflow on a call - basically, taking the crop region from the zoomed in image, then selecting the non-zoomed non-rotated original image in the same figure and clicking on Crop button.  In the crop dialog, the thumb of the Clipboard is wrong, as well as when this wrong thumb is applied onto the image, it is not where it should be either.

**Comment by pwalczysko** [comment]
The problem https://github.com/ome/omero-figure/pull/410#issuecomment-778292041 is fixed now.

Ready to merge fmpov.
